README.md
# Userscripts Safari

An open-source userscript editor for Safari - https://quoid.github.io/userscripts/

![Userscripts Safari](/etc/screenshot.png)

## Installation

Install via [Mac App Store](https://itunes.apple.com/us/app/userscripts/id1463298887) or clone the project and build with Xcode.

**If you are looking for the old/simplified version of this extension, please check out [InjectJS](https://github.com/quoid/InjectJS).**

## Usage

**Version 2 and above requires Safari 13+**

As of Version 2.0.0, this extension now supports multiple, domain specific, userscripts as well as supporting 2 different "userscript" types. It supports traditional javascript userscripts as well and css "userscripts" (or more commonly referred to as, userstyles).

Click the extension's toolbar button to open the extension UI.

**If you don't feel like reading, view [this markdown file](/etc/gifs.md) to watch some short agif tutorials.**

**UI Overview**
![Userscripts Safari](/etc/ui-overview-01.png)
![Userscripts Safari](/etc/ui-overview-02.png)

1. **Create New Userscript Button** - Click this button to create a new userscript. You will be able to choose what type of userscript you would like to create (css or js)
1. **Userscript Filter** - You can use this to filter through your existing userscript. **Note**, this element filters by userscript title, not desription. It supports the following flags to help you filter more efficiently: `!js`, `!css`, `!disabled`, `!enabled`
1. **Userscript Toggle** - Toggling this element with enable/disable it's injection into webpages
1. **Editor** - The editor; the place where you write the code for the userscripts
1. **Download Button** - Click this to download a copy of your userscript
1. **Delete Button** - Click this to delete the userscript; *note:* deleted userscripts get put into your trash bin
1. **Settings Button** - Click this to open the settings modal
1. **Status Info** - Look here for status message pertaining to the currently opened userscript
1. **Discard/Save Buttons** - After editing a userscript, clicking on the "discard" button will remove any changes made since the userscript was loaded; clicking on the "save" button will save your changes to the userscript
1. **Sidebar** - Newly added/created scripts go here
1. **Settings Modal** - The various settings for the extension
1. **Global Blacklist** - Domain patterns entered here will be excluded from userscript injection; adding `*://*.*` to the global blacklist will turn off injection for all userscripts

## Userscript Metadata

Version 2.0.0 introduces the common practice of using userscript metadata to drive certain functionality. The currently supported fields will be listed below. There *are* plans to support more fields in future updates.

- `@name` - This will be the name that displays in the sidebar and be used as the filename - you can not use the same name for multiple scripts of the same type
- `@description`- Use this to describe what your userscript does - this will be displayed in the sidebar - there is a setting to hide userscript descriptions
- `@match` - Domain match patterns - you can use several instances of this field if you'd like multiple domain matches - view [this article for more information on constructing patterns](https://developer.chrome.com/extensions/match_patterns)
- `@exclude-match` - Domain patterns where you do *not* want the script to run
- `@include` - An alias for `@match` - functions exactly like `@match`
- `@exclude` - An alias for `@exclude-match` - functions exactly like `@exclude-match`
- `@inject-into` - allows the user to choose which context to inject the script into
    - allows the user to choose which context to inject the script into
    - values: auto, content, page (default)
    - works like [violentmonkey](https://violentmonkey.github.io/api/metadata-block/#inject-into)
- `@run-at`
    - allows the user to choose the injection timing
    - document-start, document-end (default), document-idle
    - works like [violentmonkey](https://violentmonkey.github.io/api/metadata-block/#run-at)
- `@weight`
    - allows the user to further adjust script injection timing
    - can be used to ensure one script injects before another
    - ONLY accepts integers (floats, strings and everything else will be ignored)
    - min value = 1, max value = 999, higher numbers (“heavier”) execute earlier

**All userscripts need at least 1 `@match` or `@include` to run!**

CSS Userscripts/styles can use the javascript metadata format or the Userstyle format. Both formats shown below are valid for **css** type.

```js
// ==UserScript==
// @name           Example
// @description    An example
// @match          *://github.com/*
// @match          *://*.github.com/*
// ==/UserScript==
```

```css
/* ==UserStyle==
@name           Example
@description    An example
@match          *://github.com/*
@match          *://*.github.com/*
==/UserStyle== */
```

## Userscript Save Location

See how it is done [here](/etc/gifs.md#change-save-location).

As of v2.1.0, the user can choose where their userscripts are saved to and loaded from. There's no api that allows this change to happen from the extension, so users must perform location changes from the host application.

In order to change the location, you must either launch `Userscripts.app` as you would any other application, or open the settings modal within the browser extension and click the "settings" icon within the "Save location" row.

Once the host app is open, you will see a button called "Change save location". Click this and select the directory where you'd like to save to/load from.

**Save Location Notes**

- Close all instances of the extension UI (browser app) before you change the save location. The extension will attempt to do this automatically, but it could fail (if it does, please bug report)

- Currently, when changing the save location, the app does not copy over userscripts from the previous save/load directory

- After a new save location is selected, if you rename or move that selected folder, the extension will continue to load/save to that location - the only way to remove the “link” is by trashing the folder or selecting a new save location

## Notes & FAQs

**Do I need to use the extension's editor to create new userscripts or to edit existing?**

No; simply put your userscript in the extension's scripts folder. You can easily access this folder by clicking the link in the extension's settings modal option labelled "Save Location" or by visiting `~/Library/Containers/com.userscripts.macos.Userscripts-Extension/Data/Documents/scripts/`.

If you are adding a new userscript, before injection will occur, you must open the extension UI by clicking on the extension's toolbar button. If you are editing an existing userscript through your own editor, you should not need to open the ui for the changes to take effect.

**Can I change the location where my files are saved?**

Not yet, but this is planned in a future update.

**Are there any keyboard shortcuts?**

Currently there is only one keyboard shortcut - `cmd+s` to save changes. *Note:* normal shortcuts like `cmd+z`, `cmd+shift+z`, etc... like you would have on any other normal webpage should still function.

**I updated from Version 1.5.0, where did my previous userscript go?**

Your previous userscript file *should* still exist after update. You can access its folder by clicking the link in the extension's settings modal option labelled "Save Location" or by visiting `~/Library/Containers/com.userscripts.macos.Userscripts-Extension/Data/Documents/`.

**How can I help development**

One of the best ways you can help development is to **[sign up to be a beta tester](https://forms.gle/QB46uYQHVyCxULue9)**. You will be able to test new features before they are released and help guide development by providing essential feedback.

Another great way you can help out is by **[leaving a review for the app on the App Store](https://itunes.apple.com/us/app/userscripts/id1463298887)**.

Further, any issue marked "help wanted" is actively seeking assistance. Please respond to those issues with feedback, guidance or offer coding assistance. Outside of those issues, it is fair to assume outstanding issues are already being worked on and new features are being actively developed.

## Feedback & Support
If you need help doing something with this extension or you have an idea for an enhancement, please open an issue. Be sure to search current and closed issues before opening a new one.

## Donate

## Privacy Policy
Userscripts does not collect any data from its users nor monitor activities or actions you perform within the application and extension. This means everything that you do with the application and extension is private to you and is never shared with the developers or third parties. Since there is no data collection, there is no data retention of any kind.

## License

Licensed under the [GNU General Public License v3.0](/LICENSE) license for all open source applications. A commercial license is required for all commercial applications.





CHANGELOG.md
# Changelog
## v2.2.1 (2020-12-06)
- Build for Apple silicon

## v2.2.0 (2020-10-12)
- In editor search
    - click on the search icon or hit cmd+f to bring up the search bar
    - look here for more info: https://codemirror.net/demo/search.html
- New metadata keys
    - `@inject-into`
        - allows the user to choose which context to inject the script into
        - values: auto, content, page (default)
        - works like violentmonkey - https://violentmonkey.github.io/api/metadata-block/#inject-into
    - `@run-at`
        - allows the user to choose the injection timing
        - document-start, document-end (default), document-idle
        - works like violentmonkey - https://violentmonkey.github.io/api/metadata-block/#run-at
    - `@weight`
        - allows the user to further adjust script injection timing
        - can be used to ensure one script injects before another
        - ONLY accepts integers (floats, strings and everything else will be ignored)
        - min value = 1, max value = 999, higher numbers (“heavier”) execute earlier

## v2.1.3 (2020-09-17)
- fix strict csp fallback method bug caused by safari 14 update #75

## v2.1.2 (2020-09-16)
- fixed a bug that caused discard & save buttons to remain active after loading a new script
- update styling code
- improve web demo experience

## v2.1.1 (2020-09-14)
- fixed an issue that caused extension to crash when text preceded opening userscript tag - #69 #68 #58
- improve sidebar tag filtering
- update style overflows - #71

## v2.1.0 (2020-09-08)
- allow users to choose the desired save/load location for their userscripts - #32
- refocus editor after saving - #60
- fixed a bug that was causing duplicate entries in the manifest - #62
- regularly purge manifest of stale entries - #57
- open save/load location properly when clicking link in settings modal
- misc styling/code changes

## v2.0.1 (2020-07-27)
- @include/@exclude are now aliases for @match/@exlcude-match
- Display an error when browser is not supported

## v2.0.0 (2020-07-10)
- Complete overhaul/rewrite of the extension
- UI now lives in a page rather than popover
- Enables multiple, domain specific, scripts
- A lot more features of which will not be listed in the changelog

## v1.5.0 (2019-10-14)
- Enable code folding in editor
- Minor styling changes

## v1.4.0 (2019-08-28)
- A lot of code cleanup and refactoring
- Improved the popover loading experience
- Javascript is now saved to a .js file, rather than .json

## v1.3.0 (2019-07-24)
- Refactored WKWebView implementation to better support Catalina
- View styling changes to prevent initial "white flash" before WKWebView loads
- Minor CSS style updates

## v1.2.0 (2019-07-02)
- Fixed an issue that prevented script injection on websites with strict CSP
- Disabled 3D touch within the editor, which was causing cursor issues
- Styling updates

## v1.1.0 (2019-06-13)
- Added the ability for users to toggle the extension on/off
- Updated the popover dimensions
- Tabs are now transformed into spaces within the editor
- Created a workaround so that the mouse cursor changes appropriately when hovering certain elements
- Fixed a bug that allowed users to overwrite a saved script when no changes occurred
- Fixed a bug that was showing a undefined "Last edited" date on first launch of the extension

## v1.0.3 (2019-05-13)

Initial release using [Safari App Extensions](https://developer.apple.com/documentation/safariservices/safari_app_extensions) format

## v1.0.0 (2018-12-01)

Initial release using [Safari Extension JS](https://developer.apple.com/documentation/safariextensions) - depreciated as of 2019-01-01


src/js/dev.js
"use strict";

const ___safari = {
    delay: 500,
    allScriptsData: function() {
        // iterate through all scripts in save location
        // ensure each has proper extension
        // validate script content
        // return *array* of individualized script data
        const arr = [];
        for (const file in ___files) {
            const ext = ___h.getFileExtension(file);
            if (ext === "js" || ext === "css") {
                const content = ___files[file]["content"];
                const parsed = ___parse.run(content);
                if (!___files[file] || !content || !parsed ) {
                    continue;
                }
                const metadata = parsed["metadata"];
                if (!metadata["name"]) {
                    continue;
                }
                const scriptData = {
                    disabled: ___manifest["disabled"].includes(file),
                    filename: file,
                    lastModifiedMS: ___files[file]["lastModifiedMS"],
                    name: metadata["name"][0],
                    type: ___h.getFileExtension(file)
                };
                if (metadata["description"]) {
                    scriptData["description"] = metadata["description"][0];
                }
                arr.push(scriptData);
            }
        }
        return arr;
    },
    dispatchMessageToScript: function(name, data) { // arg1 string, arg2 any
        const event = {
            name: name,
            message: data
        };
        setTimeout(function() {
            this.log("MessageReceived:", name, "green");
            ___a.handleMessage(event);
        }.bind(this), this.delay);
    },
    log: function(prefix, name, color) {
        const style = `color: ${color}; font-weight: bold`;
        console.log(`%c${prefix} %c${name}`, style, "font-style: normal");
    },
    saveToFiles: function(content, lastMod, newName, oldName) {
        delete ___files[oldName];
        ___files[newName] = {
            content: content,
            lastModified: lastMod
        };
    },
    singleScriptData: function(data) { // arg1 obj {id}
        // return single script data by id
        const filename = data["id"];
        const content = ___files[filename]["content"];
        const lastModified = ___files[filename]["lastModified"];
        const name = ___parse.run(content)["metadata"]["name"][0];
        const obj = {
            content: content,
            filename: filename,
            lastModified: lastModified,
            name: name,
            type: ___h.getFileExtension(filename)
        };
        return obj;
    },
    validateScript: function(scriptData) { // arg1 obj {content, id, type}
        const content = scriptData["content"];
        const oldFilename = scriptData["id"];
        const type = scriptData["type"];
        const parsed = ___parse.run(content);
        const fail = {error: "script validation failed"};
        // parser return null when metablock missing
        // saving requires name metafield
        if (!parsed || !parsed["metadata"]["name"]) {
            return fail;
        }
        const desc = parsed["metadata"]["description"];
        const lastModified = new Date();
        const name = parsed["metadata"]["name"][0];
        const newFilename = `${name}.${type}`;
        const success = {
            data: {
                content: content,
                description: desc,
                id: newFilename,
                lastModified: lastModified,
                name: name
            }
        };
        // if names are the same, can save overwrite file without issue
        if (newFilename === oldFilename) {
            this.saveToFiles(content, lastModified, newFilename, oldFilename);
            return success;
        }
        // not overwriting, check if filename for that type is taken
        if (Object.keys(___files).includes(newFilename)) {
            return fail;
        }
        // filename length too long
        if (newFilename.length > 255) {
            return fail;
        }
        this.saveToFiles(content, lastModified, newFilename, oldFilename);
        return success;
    },
    extension: {
        dispatchMessage: function(name, data) { // arg1 string, arg2 obj
            ___safari.log("MessageDispatched:", name, "darkorange");
            const resp = {requestOrigin: name};
            let m;
            switch(name) {
                case "REQ_ALL_SCRIPTS":
                    resp["data"] = ___safari.allScriptsData();
                    ___safari.dispatchMessageToScript("RESP_ALL_SCRIPTS", resp);
                    break;
                case "REQ_BLACKLIST_SAVE":
                    m = "RESP_BLACKLIST_SAVE";
                    resp["data"] = {"blacklist": data["blacklist"]};
                    ___safari.dispatchMessageToScript(m, resp);
                    break;
                case "REQ_DISABLE_SCRIPT":
                    resp["data"] = {"id": data["id"]};
                    m = "RESP_DISABLE_SCRIPT";
                    // resp["error"] = "can't disable script";
                    ___safari.dispatchMessageToScript(m, resp);
                    break;
                case "REQ_ENABLE_SCRIPT":
                    resp["data"] = {"id": data["id"]};
                    m = "RESP_ENABLE_SCRIPT";
                    ___safari.dispatchMessageToScript(m, resp);
                    break;
                case "REQ_INIT_DATA":
                    m = "RESP_INIT_DATA";
                    resp["data"] = ___settings;
                    setTimeout(function() {
                        ___safari.dispatchMessageToScript(m, resp);
                    }, this.delay);
                    break;
                case "REQ_SAVE_LOCATION_CHANGE":
                    alert("This will open the finder location when running app");
                    m = "RESP_SAVE_LOCATION_CHANGE";
                    ___safari.dispatchMessageToScript(m, resp);
                    break;
                case "REQ_SCRIPT_DELETE":
                    m = "RESP_SCRIPT_DELETE";
                    delete ___files[data["id"]];
                    resp["data"] = data;
                    // resp["error"] = "can't delete file";
                    ___safari.dispatchMessageToScript(m, resp);
                    break;
                case "REQ_SCRIPT_SAVE":
                    const responseObject = ___safari.validateScript(data);
                    const key = Object.keys(responseObject)[0];
                    resp[key] = responseObject[key];
                    // resp["error"] = "can't save file";
                    ___safari.dispatchMessageToScript("RESP_SCRIPT_SAVE", resp);
                    break;
                case "REQ_SETTING_CHANGE":
                    ___settings[data["id"]] = data["value"];
                    resp["data"] = data;
                    // resp["error"] = "can't change setting";
                    m = "RESP_SETTING_CHANGE";
                    ___safari.dispatchMessageToScript(m, resp);
                    break;
                case "REQ_SINGLE_SCRIPT":
                    resp["data"] = ___safari.singleScriptData(data);
                    //resp["error"] = "can't find script";
                    m = "RESP_SINGLE_SCRIPT";
                    ___safari.dispatchMessageToScript(m, resp);
                    break;
                default:
                    console.error(`No response match for ${name}`);
            }
        }
    }
};

var safari = ___safari;

const ___settings = {
    autoShowHints: "true",
    blacklist: [],
    hideDescriptions: "false",
    languageCode: "en-GB",
    lint: "false",
    saveLocation: "/Users/UserName/Library/Containers/ParentFolder/SubFolder/",
    showInvisibles: "true",
    tabSize: "4",
    verbose: "true",
    version: "2.2.1"
};

const ___parse = {
    run: function(str) {
        const blocksReg = /\B(?:(\/\/ ==UserScript==\r?\n([\S\s]*?)\r?\n\/\/ ==\/UserScript==)([\S\s]*)|(\/\* ==UserStyle==\r?\n([\S\s]*?)\r?\n==\/UserStyle== \*\/)([\S\s]*))/;
        const blocks = str.match(blocksReg);

        if (!blocks) {
            return null;
        }

        var metablock = (blocks[1] != null) ? blocks[1] : blocks[4];
        var metas = (blocks[2] != null) ? blocks[2] : blocks[5];
        var code = (blocks[3] != null) ? blocks[3].trim() : blocks[6].trim();

        var metadata = {};
        var metaArray = metas.split('\n');

        metaArray.forEach(function(m) {
            //  -> // @([\w-]+)[ \t]+([^\s]+[^\r\n\t\v\f]*)
            var parts = m.match(/@([\w-]+)\s+(.+)/);
            if (parts) {
                metadata[parts[1]] = metadata[parts[1]] || [];
                metadata[parts[1]].push(parts[2]);
            }
        });
        // fail if @name is missing
        if (!metadata["name"]) {
            return;
        }
        return {
            code: code,
            content: str,
            metablock: metablock,
            metadata: metadata
        }
    }
};

const ___exampleUserscriptContent = `// ==UserScript==
// @name        Example Userscript
// @description This an example userscript. It shows all the supported metablocks keys and values at this time. Since you are viewing this in the demo or development server, it will not be injected into any websites.
// @match       https://github.com/quoid/userscripts
// @exclude-match: https://www.website.com
// ==/UserScript==

console.log("I am an example userscript");
`;

const ___exampleUserscriptContent2 = `// ==UserScript==
// @name        Style Example
// @description This is an example of an injectable style code snippet. It uses the ==UserScript== format.
// @match       *://*.github.com
// ==/UserScript==

#id,
.class,
.pseudo-element::after {
    background-color: tomato;
}
`;

const ___exampleUserscriptContent3 = `/* ==UserStyle==
@name        Style Example 02
@description This is another style code example, however this example users a UserStyle metablock.
@match       https://userstyles.org/*
==/UserStyle== */

#id,
.class,
.pseudo-element::after {
    background-color: tomato;
}

`;

const ___files = {
    "foo.js": {
        "content": ___exampleUserscriptContent,
        "lastModified": "29 Dec 2019, 09:08",
        "lastModifiedMS": new Date("December 29, 2019 09:08:00").getTime()
    },
    "Style Example.css": {
        "content": ___exampleUserscriptContent2,
        "lastModified": "29 Dec 2019, 11:08",
        "lastModifiedMS": new Date("December 29, 2019 11:08:00").getTime()
    },
    "Style Example 02.css": {
        "content": ___exampleUserscriptContent3,
        "lastModified": "29 Dec 2019, 12:08",
        "lastModifiedMS": new Date("December 29, 2019 12:08:00").getTime()
    },
    "no_file_ext": {},
    "unsupported_file_extension.txt": {},
};

const ___manifest = {
    "blacklist": [],
    "disabled": [],
    "excludes": {},
    "includes": {
        "*://*/*": {
            "document-start": [
                "Example Userscript.js"
            ],
            "document-end": []
        }
    }
};
